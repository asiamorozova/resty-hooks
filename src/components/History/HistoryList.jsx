import React from 'react';
import PropTypes from 'prop-types';
import History from './HistoryItem';

const HistoryList = ({ history }) => {
  const historyList = history.map((element, i) => (
    <li key={i}>
      <History{... element }/>
    </li>
  ));

  return (
    <ul>
      {historyList}
    </ul>

  );
};

HistoryList.propTypes = {
  history: PropTypes.arrayOf(PropTypes.shape({
    url: PropTypes.string.isRequired,
    method: PropTypes.string.isRequired

  })).isRequired
};
