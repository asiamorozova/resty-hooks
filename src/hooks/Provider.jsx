import React, { createContext, useState, useContext } from 'react';
import { makeFetch } from '../services/makeFetch';
import { useLocalStorage } from './localStorage';

const Context = createContext;

export const RestyProvider = ({ children }) => {
  const [url, setUrl] = useState('');
  const [method, setMethod] = useState('GET');
  const [body, setBody] = useState('');
  const [response, setResponse] = useState({});
  const [history, setHistory] = useLocalStorage('history', []);

  const onChange = ({ target }) => {
    if(target.name === 'url') setUrl(target.value);
    if(target.name === 'method') setMethod(target.value);
    if(target.name === 'body') setBody(target.value);

  };

  const onSubmit = event => {
    event. preventDefault();
    makeFetch(url, method, body)
      .then(response => {
        setResponse(response);
        setHistory(prevHistory => [{ url, method }, ...prevHistory]);
      });
  };

  return (
    <Context.Provider value={{ url, method, body, onChange, onSubmit, response, history }}>
      {children}
    </Context.Provider>
  );
};

export const useResty = () => {
  const context = useContext(Context);
  return context;
};


