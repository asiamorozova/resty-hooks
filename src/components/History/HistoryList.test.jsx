import React from 'react';
import { shallow } from 'enzyme';
import HistoryList from './HistoryList.jsx';
import { RestyProvider } from '../../hooks/Provider';

describe('RequestHistory', () => {
  it('matches a snapshot', () => {
    const history = [{
      method: 'post', 
      url: 'http://place-puppy.com/666x666'
    }, {
      method: 'get', 
      url: 'http://place-puppy.com/666x666'
    }, {
      method: 'put', 
      url: 'http://place-puppy.com/666x666'
    }, {
      method: 'put', 
      url: 'http://place-puppy.com/666x666'
    }];

    const wrapper = shallow(
      <RestyProvider>
        <HistoryList history={history} />
      </RestyProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
