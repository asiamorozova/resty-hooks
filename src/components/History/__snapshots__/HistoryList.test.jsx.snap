// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequestHistory matches a snapshot 1`] = `
<Component
  value={
    Object {
      "body": "",
      "history": Array [],
      "method": "GET",
      "onChange": [Function],
      "onSubmit": [Function],
      "response": Object {},
      "url": "",
    }
  }
>
  <[object Object]
    history={
      Array [
        Object {
          "method": "post",
          "url": "http://place-puppy.com/666x666",
        },
        Object {
          "method": "get",
          "url": "http://place-puppy.com/666x666",
        },
        Object {
          "method": "put",
          "url": "http://place-puppy.com/666x666",
        },
        Object {
          "method": "put",
          "url": "http://place-puppy.com/666x666",
        },
      ]
    }
  />
</Component>
`;
